### PR TITLE
XmonSimulator now requires given circuits to have an XmonDevice that validates the circuit

### PR DESCRIPTION
**Pull request by Strilanc**
- No longer auto decomposes incompatible circuits
- No longer allows circuits using the unconstrained device
- Fixed xmon_simulator_test breaking these rules constantly
- Benchmarking code is now careful to generate local operations

Part of https://github.com/quantumlib/Cirq/issues/1386

### CODE REVIEW

File: cirq/circuits/circuit.py [hunk 1]
```diff
@@ -171,7 +171,7 @@ def __getitem__(self, key: int) -> ops.Moment:
 
     def __getitem__(self, key):
         if isinstance(key, slice):
-            return Circuit(self._moments[key])
+            return Circuit(self._moments[key], self.device)
```
Comment (dabacon) [inline cirq/circuits/circuit.py]: should maybe add test to check that device is propagated

File: cirq/google/sim/xmon_simulator.py [hunk 2]
```diff
@@ -133,15 +143,20 @@ def _run(
         repetitions: int,
     ) -> Dict[str, List[np.ndarray]]:
         """See definition in `cirq.SimulatesSamples`."""
-        xmon_circuit, keys = self._to_xmon_circuit(
-            circuit,
-            param_resolver)
-        if xmon_circuit.are_all_measurements_terminal():
-            return self._run_sweep_sample(xmon_circuit, repetitions)
+
```
Comment (dabacon) [inline cirq/google/sim/xmon_simulator.py]: update doc describing the verification and throwing of error

File: cirq/circuits/circuit.py (hunk 1 above)
Comment (Strilanc) [reply cirq/circuits/circuit.py]: Done.

File: cirq/google/sim/xmon_simulator.py (hunk 2 above)
Comment (Strilanc) [reply cirq/google/sim/xmon_simulator.py]: Done.

**Pull request merged**